be/src/connector/jdbc_connector.cpp

```diff
@@ -0,0 +1,123 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+#include "connector/jdbc_connector.h"
+
+#include <sstream>
+
+#include "exec/exec_node.h"
+#include "exec/vectorized/jdbc_scanner.h"
+#include "exprs/expr.h"
+#include "runtime/jdbc_driver_manager.h"
+#include "storage/chunk_helper.h"
+
+namespace starrocks {
+namespace connector {
+using namespace vectorized;
+
+// ================================
+
+DataSourceProviderPtr JDBCConnector::create_data_source_provider(vectorized::ConnectorScanNode* scan_node,
+                                                                 const TPlanNode& plan_node) const {
+    return std::make_unique<JDBCDataSourceProvider>(scan_node, plan_node);
+}
+
+// ================================
+
+JDBCDataSourceProvider::JDBCDataSourceProvider(vectorized::ConnectorScanNode* scan_node, const TPlanNode& plan_node)
+        : _scan_node(scan_node), _jdbc_scan_node(plan_node.jdbc_scan_node) {}
+
+DataSourcePtr JDBCDataSourceProvider::create_data_source(const TScanRange& scan_range) {
+    return std::make_unique<JDBCDataSource>(this, scan_range);
+}
+
+// ================================
+
+static std::string get_jdbc_sql(const std::string& table, const std::vector<std::string>& columns,
+                                const std::vector<std::string>& filters, int64_t limit) {
+    std::ostringstream oss;
+    oss << "SELECT";
+    for (size_t i = 0; i < columns.size(); i++) {
+        oss << (i == 0 ? "" : ",") << " " << columns[i];
+    }
+    oss << " FROM " << table;
+    if (!filters.empty()) {
+        oss << " WHERE ";
+        for (size_t i = 0; i < filters.size(); i++) {
+            oss << (i == 0 ? "" : " AND") << "(" << filters[i] << ")";
+        }
+    }
+    if (limit != -1) {
+        oss << " LIMIT " << limit;
+    }
+    return oss.str();
+}
+
+JDBCDataSource::JDBCDataSource(const JDBCDataSourceProvider* provider, const TScanRange& scan_range)
+        : _provider(provider) {}
+
+Status JDBCDataSource::open(RuntimeState* state) {
+    const TJDBCScanNode& jdbc_scan_node = _provider->_jdbc_scan_node;
+    _runtime_state = state;
+    _tuple_desc = state->desc_tbl().get_tuple_descriptor(jdbc_scan_node.tuple_id);
+    RETURN_IF_ERROR(_create_scanner(state));
+    return Status::OK();
+}
+
+void JDBCDataSource::close(RuntimeState* state) {
+    if (_scanner != nullptr) {
+        _scanner->close(state);
+    }
+}
+
+Status JDBCDataSource::get_next(RuntimeState* state, vectorized::ChunkPtr* chunk) {
+    bool eos = false;
+    do {
+        RETURN_IF_ERROR(_scanner->get_next(state, chunk, &eos));
```

Comment:
we do not consider limit now, am i right?

Reply:
Because framework has done that. 

```
Status ConnectorChunkSource::_read_chunk(vectorized::ChunkPtr* chunk) {
    RuntimeState* state = _runtime_state;
    if (state->is_cancelled()) {
        return Status::Cancelled("canceled state");
    }

    // Improve for select * from table limit x, x is small
    if (_limit != -1 && _rows_read >= _limit) {
        return Status::EndOfFile("limit reach");
    }

    do {
        RETURN_IF_ERROR(_data_source->get_next(state, chunk));
    } while ((*chunk)->num_rows() == 0);
    _rows_read += (*chunk)->num_rows();

    return Status::OK();
}
```